### PR TITLE
[Kelley] add sender constraints

### PR DESCRIPTION
**Pull request by arndt-s**

### CODE REVIEW

**Review by PieterKas — APPROVED** [review]

**Review by PieterKas — CHANGES_REQUESTED** [review]
There appears to be issues on lines 140,142 163 and 241 (I suspect trailing whitespaces). @arndt-s can you update the PR with these changes (I don't seem to be able to edit it).

File: draft-ietf-oauth-identity-chaining.md [hunk 1]
```diff
@@ -131,8 +132,15 @@ The flow illustrated in Figure 1 shows the steps the client in trust Domain A ne
 * (F) The client now possesses an access token to access the protected resource in Domain B.
 
 ## Authorization Server Discovery
+
 This specification does not define authorization server discovery. A client MAY maintain a static mapping or use other means to identify the authorization server. The `authorization_servers` property in {{I-D.ietf-oauth-resource-metadata}} MAY be used.
 
+## Sender Constrained Tokens
+
+The authorization server MAY issue sender constrained tokens. There are currently two options for sender constrained tokens: Mutually Authenticated TLS (mTLS) {{RFC8705}} and Demonstrating Proof of Possessions (DPoP) {{RFC9449}}.
```
Comment (arndt-s) [inline draft-ietf-oauth-identity-chaining.md:140]: Is this in regards to authorization server in Domain A or B or both?

File: draft-ietf-oauth-identity-chaining.md [hunk 2]
```diff
@@ -131,8 +132,15 @@ The flow illustrated in Figure 1 shows the steps the client in trust Domain A ne
...
 
 ## Authorization Server Discovery
+
 This specification does not define authorization server discovery. A client MAY maintain a static mapping or use other means to identify the authorization server. The `authorization_servers` property in {{I-D.ietf-oauth-resource-metadata}} MAY be used.
 
+## Sender Constrained Tokens
+
+The authorization server MAY issue sender constrained tokens. There are currently two options for sender constrained tokens: Mutually Authenticated TLS (mTLS) {{RFC8705}} and Demonstrating Proof of Possessions (DPoP) {{RFC9449}}. 
+
+Additionally, there are two examples of identity and authorization chaining: resource server acting as the client and the authorization server acting as the client. Additional details of the examples are in [Examples](#examples).
```
Comment (arndt-s) [inline draft-ietf-oauth-identity-chaining.md:142]: This seems duplicated. In the overview section we already have 
```
Section [Examples](#examples) includes two additional examples that describe how this flow is used. In one example, the resource server acts as the client and in the other, the authorization server acts as the client.
```

File: draft-ietf-oauth-identity-chaining.md [hunk 3]
```diff
@@ -224,6 +238,7 @@ The authorization server MUST validate the JWT authorization grant as specified
 * The "aud" claim MUST identify the Authorization Server as a valid intended audience of the assertion using either the token endpoint as described Section 3 {{RFC7523}} or the issuer identifier as defined in Section 2 of {{RFC8414}}.
 * The authorization server SHOULD deny the request if it is not able to identify the subject.
 * Due to policy the request MAY be denied (for instance if the federation from domain A is not allowed).
+* If the authorization server issues sender constrained tokens, the authorization server MUST check the binding on inbound assertion of the Access Token Request. Additionally, the authorization server MUST verify the "cnf" claim.
```
Comment (arndt-s) [inline draft-ietf-oauth-identity-chaining.md:241]: Does this work if the resource server acts as a client? Thinking out loud:

```
T0: AS1 issues AT to Alice.
T1: Alice presents AT to R1. R1 needs to call R2 (in domain 2).
T2: performs token exchange to AS1 (fails as AT is bound).
```

File: draft-ietf-oauth-identity-chaining.md (hunk 1 above)
Comment (kburgin3) [reply draft-ietf-oauth-identity-chaining.md:140]: Both

File: draft-ietf-oauth-identity-chaining.md (hunk 2 above)
Comment (kburgin3) [reply draft-ietf-oauth-identity-chaining.md:142]: Suggest deleting this

File: draft-ietf-oauth-identity-chaining.md [hunk 4]
```diff
@@ -142,19 +150,23 @@ The client performs token exchange as defined in {{RFC8693}} with the authorizat
...
 
-resource
+"resource"
 : REQUIRED if audience is not set. URI of authorization server of targeting domain (domain B).
 
-audience
+"audience"
 : REQUIRED if resource is not set. Well known/logical name of authorization server of targeting domain (domain B).
 
+"chained_id"
```
[the remark this answers is not in the document]
Comment (kburgin3) [reply draft-ietf-oauth-identity-chaining.md:162]: I invented "chained_id" to carry the "cnf" claim. And yes, you understand correctly

[the remark this answers is not in the document]
Comment (bc-pi) [reply draft-ietf-oauth-identity-chaining.md:160]: I think I don't understand. This is in a list of parameters but is supposed to be a claim?

**Review by bc-pi — CHANGES_REQUESTED** [review]
In the related issue #86 I'd left [a comment](https://github.com/oauth-wg/oauth-identity-chaining/issues/86#issuecomment-2077999207) with a number of things that'd need to be accounted for or a least considered in adding support for sender constraining to this draft, which I don't believe have been addressed in this PR.  That aforementioned comment copied below for ease of access. 

> A rough sketch of things that likely need to be accounted for or a least considered in this effort:
>
> * checking the binding on inbound token of [Token Exchange](https://www.ietf.org/archive/id/draft-ietf-oauth-identity-chaining-01.html#name-token-exchange) request. No standard defined here. Might not even be feasible or make sense depending on deployment (i.e. the RS acting as the client doesn't have access to the bound key)
> * binding the token returned from Token Exchange. The request could be MTLS or have a DPoP header, which is kinda implied per standards. And the returned JWT could be bound with the respective `cnf` claim and method, which is maybe implied by standards but not explicitly written down anywhere (that I know of).
> * checking the binding on inbound token/assertion of the [Access Token Request](https://www.ietf.org/archive/id/draft-ietf-oauth-identity-chaining-01.html#name-access-token-request). The request could be MTLS or have a DPoP header, which is kinda implied per standards but kinda leaves much to the imagination, and checked against the `cnf` claim of the JWT authz grant. The presence of the `cnf` claim might be sufficient to say the thing is bound and suggest/require checking it (implementations I know about would just ignore it, however). But maybe a new grant type would make sense to clearly convey the intent (also using `urn:ietf:params:oauth:grant-type:jwt-bearer` for a non-bearer thing is awkward). 
> * binding the resultant access token. The request could be MTLS or have a DPoP header and I think this is covered/implied by the existing specs to say that the issued access token should/could be bound a la [DPoP Access Token Request](https://www.rfc-editor.org/rfc/rfc9449.html#name-dpop-access-token-request) and [Mutual-TLS Client Certificate-Bound Access Tokens](https://www.rfc-editor.org/rfc/rfc8705.html#name-mutual-tls-client-certifica). 
>
> The [Authorization server acting as client](https://www.ietf.org/archive/id/draft-ietf-oauth-identity-chaining-01.html#name-authorization-server-acting) flips around nearly all of the pieces above to make them unworkable. I guess it'd need separate treatment with the definition of something like a proxied "trust me" this is the cnf I need in the final access token. IIRC Kelley had some prior work toward this end somewhere but I can't seem to find it at the moment. Or maybe sender constraining is just out of scope in the AS as client proxy case.

File: draft-ietf-oauth-identity-chaining.md [hunk 5]
```diff
@@ -142,19 +148,23 @@ The client performs token exchange as defined in {{RFC8693}} with the authorizat
...
+: REQUIRED if the authorization server issues sender constrained tokens. The chained_id claim contains as a sub-claim the "cnf" value from the inbound token that was presented during token exchange.
+
 ### Processing rules
 
 * If the request itself is not valid or if the given resource or audience are unknown, or are unacceptable based on policy, the authorization server MUST deny the request.
 * The authorization server MAY add, remove or change claims. See [Claims transcription](#claims-transcription).
+* If the inbound token is sender constrained, the authorization server MUST check the binding on the inbound token.
```
Comment (bc-pi) [inline draft-ietf-oauth-identity-chaining.md:167]: I think this would preclude the RS acting as the client case

**Pull request closed without merging**